File: java_classes/FilterCriteria.java
package com.asfar.web.dto.filter;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.time.LocalDate;
import java.util.List;

/**
 * DTO pour les critères de filtrage des appartements
 */
public class FilterCriteria {

    @JsonProperty("prix_min")
    private Double prixMin;

    @JsonProperty("prix_max")
    private Double prixMax;

    @JsonProperty("date_debut")
    private LocalDate dateDebut;

    @JsonProperty("date_fin")
    private LocalDate dateFin;

    @JsonProperty("nb_lits")
    private Integer nbLits;

    @JsonProperty("nb_chambres")
    private Integer nbChambres;

    @JsonProperty("nb_douches")
    private Integer nbDouches;

    @JsonProperty("commodites")
    private List<String> commodites;

    @JsonProperty("preferences")
    private List<String> preferences;

    @JsonProperty("regles")
    private List<String> regles;

    // Constructeurs
    public FilterCriteria() {}

    public FilterCriteria(Double prixMin, Double prixMax, LocalDate dateDebut, LocalDate dateFin,
                         Integer nbLits, Integer nbChambres, Integer nbDouches,
                         List<String> commodites, List<String> preferences, List<String> regles) {
        this.prixMin = prixMin;
        this.prixMax = prixMax;
        this.dateDebut = dateDebut;
        this.dateFin = dateFin;
        this.nbLits = nbLits;
        this.nbChambres = nbChambres;
        this.nbDouches = nbDouches;
        this.commodites = commodites;
        this.preferences = preferences;
        this.regles = regles;
    }

    // Getters et Setters
    public Double getPrixMin() {
        return prixMin;
    }

    public void setPrixMin(Double prixMin) {
        this.prixMin = prixMin;
    }

    public Double getPrixMax() {
        return prixMax;
    }

    public void setPrixMax(Double prixMax) {
        this.prixMax = prixMax;
    }

    public LocalDate getDateDebut() {
        return dateDebut;
    }

    public void setDateDebut(LocalDate dateDebut) {
        this.dateDebut = dateDebut;
    }

    public LocalDate getDateFin() {
        return dateFin;
    }

    public void setDateFin(LocalDate dateFin) {
        this.dateFin = dateFin;
    }

    public Integer getNbLits() {
        return nbLits;
    }

    public void setNbLits(Integer nbLits) {
        this.nbLits = nbLits;
    }

    public Integer getNbChambres() {
        return nbChambres;
    }

    public void setNbChambres(Integer nbChambres) {
        this.nbChambres = nbChambres;
    }

    public Integer getNbDouches() {
        return nbDouches;
    }

    public void setNbDouches(Integer nbDouches) {
        this.nbDouches = nbDouches;
    }

    public List<String> getCommodites() {
        return commodites;
    }

    public void setCommodites(List<String> commodites) {
        this.commodites = commodites;
    }

    public List<String> getPreferences() {
        return preferences;
    }

    public void setPreferences(List<String> preferences) {
        this.preferences = preferences;
    }

    public List<String> getRegles() {
        return regles;
    }

    public void setRegles(List<String> regles) {
        this.regles = regles;
    }

    // Méthodes utilitaires
    public boolean hasFilters() {
        return prixMin != null || prixMax != null ||
               dateDebut != null || dateFin != null ||
               (nbLits != null && nbLits > 0) ||
               (nbChambres != null && nbChambres > 0) ||
               (nbDouches != null && nbDouches > 0) ||
               (commodites != null && !commodites.isEmpty()) ||
               (preferences != null && !preferences.isEmpty()) ||
               (regles != null && !regles.isEmpty());
    }

    @Override
    public String toString() {
        return "FilterCriteria{" +
                "prixMin=" + prixMin +
                ", prixMax=" + prixMax +
                ", dateDebut=" + dateDebut +
                ", dateFin=" + dateFin +
                ", nbLits=" + nbLits +
                ", nbChambres=" + nbChambres +
                ", nbDouches=" + nbDouches +
                ", commodites=" + commodites +
                ", preferences=" + preferences +
                ", regles=" + regles +
                '}';
    }
}

File: lib/model/filter/filter_criteria.dart
class FilterCriteria {
  double? prixMin;
  double? prixMax;
  DateTime? dateDebut;
  DateTime? dateFin;
  int? nbLits;
  int? nbChambres;
  int? nbDouches;
  List<String>? commodites;
  List<String>? preferences;
  List<String>? regles;

  FilterCriteria({
    this.prixMin,
    this.prixMax,
    this.dateDebut,
    this.dateFin,
    this.nbLits,
    this.nbChambres,
    this.nbDouches,
    this.commodites,
    this.preferences,
    this.regles,
  });

  FilterCriteria.fromJson(Map<String, dynamic> json) {
    prixMin = json['prix_min']?.toDouble();
    prixMax = json['prix_max']?.toDouble();
    dateDebut = json['date_debut'] != null ? DateTime.parse(json['date_debut']) : null;
    dateFin = json['date_fin'] != null ? DateTime.parse(json['date_fin']) : null;
    nbLits = json['nb_lits'];
    nbChambres = json['nb_chambres'];
    nbDouches = json['nb_douches'];
    commodites = json['commodites'] != null ? List<String>.from(json['commodites']) : null;
    preferences = json['preferences'] != null ? List<String>.from(json['preferences']) : null;
    regles = json['regles'] != null ? List<String>.from(json['regles']) : null;
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    if (prixMin != null) data['prix_min'] = prixMin;
    if (prixMax != null) data['prix_max'] = prixMax;
    if (dateDebut != null) data['date_debut'] = dateDebut!.toIso8601String().split('T')[0];
    if (dateFin != null) data['date_fin'] = dateFin!.toIso8601String().split('T')[0];
    if (nbLits != null) data['nb_lits'] = nbLits;
    if (nbChambres != null) data['nb_chambres'] = nbChambres;
    if (nbDouches != null) data['nb_douches'] = nbDouches;
    if (commodites != null && commodites!.isNotEmpty) data['commodites'] = commodites;
    if (preferences != null && preferences!.isNotEmpty) data['preferences'] = preferences;
    if (regles != null && regles!.isNotEmpty) data['regles'] = regles;
    return data;
  }

  bool get hasFilters {
    return prixMin != null ||
        prixMax != null ||
        dateDebut != null ||
        dateFin != null ||
        (nbLits != null && nbLits! > 0) ||
        (nbChambres != null && nbChambres! > 0) ||
        (nbDouches != null && nbDouches! > 0) ||
        (commodites != null && commodites!.isNotEmpty) ||
        (preferences != null && preferences!.isNotEmpty) ||
        (regles != null && regles!.isNotEmpty);
  }

  int get activeFiltersCount {
    int count = 0;
    if (prixMin != null || prixMax != null) count++;
    if (dateDebut != null || dateFin != null) count++;
    if (nbLits != null && nbLits! > 0) count++;
    if (nbChambres != null && nbChambres! > 0) count++;
    if (nbDouches != null && nbDouches! > 0) count++;
    if (commodites != null && commodites!.isNotEmpty) count++;
    if (preferences != null && preferences!.isNotEmpty) count++;
    if (regles != null && regles!.isNotEmpty) count++;
    return count;
  }

  FilterCriteria copyWith({
    double? prixMin,
    double? prixMax,
    DateTime? dateDebut,
    DateTime? dateFin,
    int? nbLits,
    int? nbChambres,
    int? nbDouches,
    List<String>? commodites,
    List<String>? preferences,
    List<String>? regles,
  }) {
    return FilterCriteria(
      prixMin: prixMin ?? this.prixMin,
      prixMax: prixMax ?? this.prixMax,
      dateDebut: dateDebut ?? this.dateDebut,
      dateFin: dateFin ?? this.dateFin,
      nbLits: nbLits ?? this.nbLits,
      nbChambres: nbChambres ?? this.nbChambres,
      nbDouches: nbDouches ?? this.nbDouches,
      commodites: commodites ?? this.commodites,
      preferences: preferences ?? this.preferences,
      regles: regles ?? this.regles,
    );
  }

  void clear() {
    prixMin = null;
    prixMax = null;
    dateDebut = null;
    dateFin = null;
    nbLits = null;
    nbChambres = null;
    nbDouches = null;
    commodites = null;
    preferences = null;
    regles = null;
  }

  @override
  String toString() {
    return 'FilterCriteria(prixMin: $prixMin, prixMax: $prixMax, dateDebut: $dateDebut, dateFin: $dateFin, nbLits: $nbLits, nbChambres: $nbChambres, nbDouches: $nbDouches, commodites: $commodites, preferences: $preferences, regles: $regles)';
  }
}